### PR TITLE
Implement Advanced Energy Detector

### PR DESCRIPTION
**Pull request by ghzdude**
**What:**
This PR implements the Advanced Energy Detector, which emits redstone when reaching a lower threshold and stops at a higher threshold. This PR is also based off of the Modular UI branch, so this will go there.

**Implementation Details:**
Could look over the `createWindow()` method in `CoverDetectorEnergyAdvanced.java` to make sure my usage of Modular UI is fine, as well as `update()`. If there's any formatting issues let me know. There's also no recipe for the cover right now, as well as missing a unique texture.

**Outcome:**
Adds and registers a cover `Advanced Energy Detector`.

**Additional info:**
https://user-images.githubusercontent.com/44148655/180641472-34b5eafa-1f6f-436f-8a39-956e4b166655.mp4

**Possible compatibility issue:**
I don't think this should have a compatibility issue, but if there's an issue please let me know.

### CODE REVIEW

**Comment by hjae78** [comment]
A mode specifying an amount of EU stored rather than a percentage could have some uses

**Comment by ghzdude** [comment]
I'm debating if there could be advanced versions of Fluid and Item detectors as well, though I'm not sure how useful that would be.

**Comment by ghzdude** [comment]
Advanced Energy Detector can now use EU instead of percent for detection
https://user-images.githubusercontent.com/44148655/180917535-b100177b-d3f9-49bb-995a-2e1e8e9e29c9.mp4

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityPassthroughHatchItem.java [hunk 1]
```diff
@@ -80,7 +80,7 @@ protected IItemHandlerModifiable createExportItemHandler() {
     protected IItemHandlerModifiable createImportItemHandler() {
         return new NotifiableItemStackHandler(getInventorySize(), getController(), false);
     }
-
+/*
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityPassthroughHatchItem.java:83]: Why are these commented out? Seems like irrelevant changes for the PR, plus something we don't really want to change.

File: src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+import com.cleanroommc.modularui.common.widget.textfield.TextFieldWidget;
+import gregtech.api.util.GTLog;
+import gregtech.client.renderer.texture.Textures;
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.*;
+
+
+public class CoverDetectorEnergyAdvanced extends CoverBehavior implements CoverWithUI, ITickable{
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:32]: This should also implement `IControllable` so that it can be controlled via machine controller. It will also need to check for the controllable capability.

File: src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            if (useEU) {
+                compareValue((float) energyContainer.getEnergyStored(), maxEU, minEU);
+            } else {
+                compareValue(currentStorageRatio, maxPercent, minPercent);
+            }
+            setRedstoneSignalOutput(outputAmount);
+        }
+    }
+
+    private void compareValue (float value, int maxValue, int minValue) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:90]: Would it be useful to emit differing redstone levels dependent upon the amount of difference between the read value and the max/min value of the cover?

**Review by ALongStringOfNumbers — CHANGES_REQUESTED** [review]
Could use a texture that is slightly different from the default version of the energy detector.

File: src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityPassthroughHatchItem.java (hunk 1 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/metatileentities/multi/multiblockpart/MetaTileEntityPassthroughHatchItem.java:83]: They were commented out because of bad imports preventing me from using `runClient`, but the imports have since been fixed so i'll uncomment these sections.

File: src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java (hunk 3 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:90]: I'm not sure what you mean by this, can you elaborate?

File: src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java (hunk 2 above)
Comment (ghzdude) [reply src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java]: I believe I have properly implemented `IControllable` now with the capability check.

File: src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java (hunk 3 above)
Comment (ALongStringOfNumbers) [reply src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:93]: No, I meant something like: EU Range is 0 to 2048, at 1024 EU, emit a signal strength of 8, at 512, a signal strength of 4, etc.

Basically scaling the output strength of the signal by how close the value read from the machine is to the defined minimum and maximum.

Comment (ghzdude) [reply src/main/java/gregtech/common/covers/detector/CoverDetectorEnergyAdvanced.java:94]: Ah, I see what you mean now. 

I think I've implemented this feature now, behind a toggle button in case it's undesirable for the player.

**Comment by LAGIdiot** [comment]
This was superseded by #1222 and can be closed.
